### PR TITLE
KAFKA-14015: StandaloneHerder::restartConnector should reconfigure tasks if configs have been changed

### PR DESCRIPTION
**Pull request by yashmayya**
- https://issues.apache.org/jira/browse/KAFKA-14015: A config provider (https://cwiki.apache.org/confluence/display/KAFKA/KIP-297%3A+Externalizing+Secrets+for+Connect+Configurations) that has dynamic reloading of secrets doesn't work as expected when Connect is run in standalone mode - i.e. tasks continue running with an older secret config value when the secret is rotated.
- When a config provider needs to inform the herder that a connector needs to be restarted due to a change in a secret config that needs to be propagated to the connector and its tasks - it calls `Herder::restartConnector`
- In the `DistributedHerder` implementation of `restartConnector` this works as expected where the connector is stopped, then started, and then if it's in the `STARTED` state at the end of startup - a task reconfiguration is requested. A task reconfiguration involves asking the connector to return a list of task configs (using the latest connector configs), comparing it with the running task configs and if a change is detected - write the new set of task configs to the config topic which will eventually result in the old set of tasks being stopped and a new set of tasks being brought up. 
- In the `StandaloneHerder` implementation, however, the step of requesting task reconfiguration is missing. This PR aims to fix this which should solve the issue in https://issues.apache.org/jira/browse/KAFKA-14015

### CODE REVIEW

**Comment by yashmayya** [comment]
@C0urante @vamossagar12 could you please take a look at this whenever you get a chance? Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java [hunk 1]
```diff
@@ -294,7 +294,12 @@ public synchronized void restartConnector(String connName, Callback<Void> cb) {
 
         worker.stopAndAwaitConnector(connName);
 
-        startConnector(connName, (error, result) -> cb.onCompletion(error, null));
+        startConnector(connName, (error, targetState) -> {
+            if (targetState == TargetState.STARTED) {
+                requestTaskReconfiguration(connName);
```
Comment (yashmayya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:299]: I'm wondering if this should be submitted as a runnable to `requestExecutorService` instead - since `restartConnector` is also called by the `POST /{connector}/restart` API, and this change could lead to a significantly higher response latency. Also, if we were to do that, should we add a `sleep` call to the relevant unit tests and retain the same expectations / verifications (with the assumption that the requestExecutorService will complete the task in the background)?

Edit: I guess a better option on the testing front might be to pass an implementation of `ScheduledExecutorService` to `StandaloneHerder` which runs tasks in the same thread.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:299]: I think this is fine as-is, we synchronize in `requestTaskReconfiguration` either way. I can't think of a significant difference between invoking that method in the `WorkerConnector`'s thread and invoking it via the herder's `requestExecutorService`.

In fact, I don't think we need to invoke anything via the `requestExecutorService` except in the variant of `restartConnector` that accepts a delay. But that's out of scope for this PR and should be addressed separately in order to simplify review.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java [hunk 2]
```diff
@@ -812,7 +861,7 @@ public void testPutConnectorConfig() throws Exception {
         FutureCallback<Herder.Created<ConnectorInfo>> reconfigureCallback = new FutureCallback<>();
         herder.putConnectorConfig(CONNECTOR_NAME, newConnConfig, true, reconfigureCallback);
         Herder.Created<ConnectorInfo> newConnectorInfo = reconfigureCallback.get(1000L, TimeUnit.SECONDS);
-        ConnectorInfo newConnInfo = new ConnectorInfo(CONNECTOR_NAME, newConnConfig, Arrays.asList(new ConnectorTaskId(CONNECTOR_NAME, 0)),
+        ConnectorInfo newConnInfo = new ConnectorInfo(CONNECTOR_NAME, newConnConfig, singletonList(new ConnectorTaskId(CONNECTOR_NAME, 0)),
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java]: My IDE has been nagging me about this one too, for years. I've left it as-is (maybe not here but in other places) since it's not performance-critical and it's easier to modify in case we need to add other elements, or construct an empty list. I think we should revert this change.

**Review by C0urante — COMMENTED** [review]
Looking great, one small comment and then this should be good to go. Thanks Yash!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java (hunk 2 above)
Comment (yashmayya) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java]: Haha yeah I only did this so that my IDE stopped nagging me 😄 

>it's easier to modify in case we need to add other elements, or construct an empty list

Fair point, I've reverted this.

**Review by C0urante — APPROVED** [review]
LGTM
